### ISSUE TITLE
(Suppression of) intercept term in lhs of random effects

### ISSUE THREAD

**Issue by dmbates** [dmbates/MixedModels.jl#195]
I just saw that with a experimental factor on the left hand side of random-effects term, suppressing the intercept does not give the desired model.

I can fit a model with an experimental factor, `f`, in the random effects specification for the grouping factor, `g`, and get the expected result.
```julia
julia> using DataFrames, MixedModels

julia> dat = DataFrame(y = ones(18),
            g = string.(repeat('a':'f', inner=3)),
            f = string.(repeat('A':'C', outer=6)));

julia> mod = LinearMixedModel(@formula(y ~ 1 + f + (1+f|g)), dat);

julia> refit!(simulate!(mod, β = [1.0, -0.2, 0.3], σ = 0.05,
                        θ = [1.0, -0.1, 0.3, 0.2, 0.1, 0.3]))
Linear mixed model fit by maximum likelihood
 y ~ 1 + f + (1 + f | g)
   logLik   -2 logLik     AIC        BIC    
  24.297916 -48.595831 -28.595831 -19.692114

Variance components:
              Column      Variance     Std.Dev.    Corr.
 g        (Intercept)  0.00713718070 0.084481836
          f: B         0.00060249946 0.024545864  0.54
          f: C         0.00282675031 0.053167192  0.35 -0.48
 Residual              0.00067137974 0.025910997
 Number of obs: 18; levels of grouping factors: 6

  Fixed-effects parameters:
────────────────────────────────────────────────────
              Estimate  Std.Error   z value  P(>|z|)
────────────────────────────────────────────────────
(Intercept)   0.994279  0.0360753   27.5612   <1e-99
f: B         -0.221398  0.0180058  -12.2959   <1e-34
f: C          0.351254  0.0263613   13.3246   <1e-39
────────────────────────────────────────────────────

julia> dat.y = mod.y;

julia> mod0 = fit(MixedModel, @formula(y ~ 0 + f + (0+f|g)), dat)
Linear mixed model fit by maximum likelihood
 y ~ 0 + f + (0 + f | g)
   logLik   -2 logLik     AIC        BIC    
  17.362429 -34.724857 -20.724857 -14.492255

Variance components:
           Column   Variance   Std.Dev.    Corr.
 g        f: B    0.006736238 0.08207459
          f: C    0.009447073 0.09719606  1.00
 Residual         0.005346109 0.07311709
 Number of obs: 18; levels of grouping factors: 6

  Fixed-effects parameters:
───────────────────────────────────────────
      Estimate  Std.Error  z value  P(>|z|)
───────────────────────────────────────────
f: A  0.994279  0.0298499  33.3092   <1e-99
f: B  0.77288   0.0448745  17.2231   <1e-65
f: C  1.34553   0.0496541  27.0981   <1e-99
───────────────────────────────────────────
```
Notice that the vector-valued random effects in the second model, using a term of the form `(0+f|g)`, are two-dimensional, not three-dimensional like β.

Something in `src/randomeffectsterm.jl` needs to be modified in the way the modelcols for the lhs is evaluated.  I may need help from @kleinschmidt on this one.

**Comment by kleinschmidt** [comment]
The issue is that the `FullRank` schema is tracking things that were seen when parsing the fixed effects schema.  The solution to this is to just re-wrap the schema by adding
```julia
    # create a new schema wrapper
    schema = StatsModels.FullRank(schema.schema)
```
in the `apply_schema` method.

However, there's a deeper issue which is that the implicit intercept behavior is not handled properly for random effects, since it's handled in the `apply_schema(::FormulaTerm, ...)` method:

https://github.com/JuliaStats/StatsModels.jl/blob/d03b5c2aa64219fa63a9d43a93be5d9db221b530/src/schema.jl#L258-L277

Fixing this requires either duplicating some of the code from that method, *or* doing a dirty cheat and pretending that you're parsing a formula (since the LHS of the ranef behaves kind of like the RHS of a formula), doing something like `apply_schema(t.args_parsed[2] ~ t.args_parsed[1], schema, Mod)` and then pulling out the lhs and rhs returned there and plugging them into the `RandomEffectsTerm`...but that seems like a really dirty hack.

**Comment by kleinschmidt** [comment]
As an example of this problem:

```julia
julia> ff = apply_schema(@formula(y ~ f + (f | g)), schema(dat), MixedModel)

FormulaTerm
Response:
  y(continuous)
Predictors:
  1
  f(DummyCoding:3→2)
  (f | g)
```
(note that the "implicit intercept" term is present for the fixed effects but not the random effects)

But using this method for `apply_schema`:
```julia
function StatsModels.apply_schema(t::FunctionTerm{typeof(|)}, schema::StatsModels.FullRank,
                                  Mod::Type{<:MixedModel})
    # create a new schema wrapper
    schema = StatsModels.FullRank(schema.schema)
    lhs, rhs = t.args_parsed
    fake_formula = apply_schema(rhs ~ lhs, schema, Mod)
    RandomEffectsTerm(fake_formula.rhs, fake_formula.lhs)
end
```

You get the correct/expected result:

```julia
julia> ff = apply_schema(@formula(y ~ f + (f | g)), schema(dat), MixedModel)

FormulaTerm
Response:
  y(continuous)
Predictors:
  1
  f(DummyCoding:3→2)
  (1 + f | g)
```

**Comment by kleinschmidt** [comment]
Like we discussed in person, this might not actually be a big issue, if you want to treat implicit vs. explicit intercepts differently in a random effects term.  For instance, because of the regularization in how the random effects are estimated, it's okay to have "overcomplete" contrast coding (`FullDummyCoding` even when an intercept is present).

**Comment by palday** [comment]
I was thinking about this too.

Maybe have explicit intercepts work the way they always do. That's obvious 
We then have two bits of available syntax: no-intecerpt `(0+f|g)` and implicit intercept `(f|g)`. In `lme4`, the no-intercept formulation was typically used for achieving the zero-correlation structure by creating copies of the grouping variable.:

```r
R> lmer(Reaction ~ 1 + Days + (1|Subject) + (0+Days|Subject), sleepstudy)
Linear mixed model fit by REML ['lmerMod']
Formula: Reaction ~ 1 + Days + (1 | Subject) + (0 + Days | Subject)
   Data: sleepstudy
REML criterion at convergence: 1743.669
Random effects:
 Groups    Name        Std.Dev.
 Subject   (Intercept) 25.050  
 Subject.1 Days         5.989  
 Residual              25.565  
Number of obs: 180, groups:  Subject, 18
Fixed Effects:
(Intercept)         Days  
     251.41        10.47  
```
The double-bar syntax just re-wrote the formula to use this trick, which also explains why it didn't suppress the correlations between levels of a factor. In MixedModels, we now have `zerocorr!` for this and it does suppress all correlations within a given blocking variable. If the user wants some, but not the others, then they are free to create a duplicate blocking variable and chunk the RE as they see fit, but anything with the same blocking variable will be grouped together internally (so I thought...), so the lme4 formula trick won't work:
```julia
julia> fit!(LinearMixedModel(@formula(Reaction ~ 1 + Days + (1|Subject) + (0+Days|Subject)), rcopy(R"sleepstudy"))) # currently gives a method error
```
How much would be 'break'/shock users if  we interpret the implicit and explicit intercept formulations differently? My suggestion would be:

`(1+f|g)`: intercept + whatever contrasts f has
`(0+f|g)`: no intercept and full dummy / indicator / overparameterized matrix for f
`(f|g)`: no intercept and whatever contrasts f has *even if these don't make sense without an intercept*

In other words: the intercept can never be implicit in the random effects. For continuous `f`, `(0+f|g)` should be interpreted the same as `(f|g)`.

This depends on us being able to lump different RE terms with the same grouping variable together. For example, if we had an additional continuous covariate `u`, we could consider `(0+f|g) + (1+u|g)`   as being the overparameterized random effects for `f` and the 'normal' random effects for the intercept and `u`. Repeated terms are of course eliminated: `(1+f|g) + (1+u|g)` yields the same model as `(1+f+u|g)`.

For interaction terms in overparametrized RE, everything works the same way, just with indicators for every level of the Cartesian product of `f1` and `f2`.

**Comment by kleinschmidt** [comment]
Just to summarize our discussion offline: we can accomplish everything we need using the standard interpretation of the LHS terms, as long as we also provide a way to force "full-rank dummy coding" for some terms (in order to generate "overcomplete" random effects, where the intercept AND full-rank dummy coding is included at the same time).  Then we'd have the following (for categorical `f` with `k` levels):

* `(f | g)` - implicit intercept, default contrasts for `f` (1 intercept col + k-1 for `f`)
* `(1 + f | g)` - explicit intercept, same as implicit
* `(0 + f | g)` - no intercept, full-rank dummy for `f` (k columns)
* `(1 + FullRankDummyCoding(f) | g)` - overcomplete (1 intercept  + k for `f`)

I think implementing support for specifying contrasts inside the formula like this is best handled in statsmodels.

**Comment by palday** [comment]
Yep, and then the behavior is consistent with the RHS in the fixed effects, which has two advantages:

1. The fix for the original issue remains unchanged.
2. This is consistent behavior because the LHS of the random effects really is just a by-`g` RHS.
